### PR TITLE
Add azure managed identity to azure package in blockstorage

### PR DESCRIPTION
**Pull request by leuyentran**
## Change Overview

This PR add : 

1.  Support for Azure Manage Identity by adding MSI authorizer
2. Add an Authenticate() interface to check validity of Azure creds.  Azure Managed Identity is  the first type to be supported for this interface .  Similar to other Azure interfaces, this interface is not united tested because creating a new Azure provider involves calling `Instance Azure Metadata` which will results in `context timeout `
4. Add utilities functions to provide appropriate abstractions levels for point 1 and 2 + unit tests

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [x] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/blockstorage/azure/azuredisk.go
Comment (leuyentran) [inline pkg/blockstorage/azure/azuredisk.go:48]: This interface is not united tested because creating a new Azure provider involves calling `Azure Instance Metadata` which will results in `context timeout`

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [inline pkg/blockstorage/azure/auth.go:65]: This interface is not united tested because it involves network calls which will results in `context timeout`

File: pkg/blockstorage/azure/auth.go
Comment (bathina2) [inline pkg/blockstorage/azure/auth.go:22]: Is this setting MSI as default?

Comment (bathina2) [reply pkg/blockstorage/azure/auth.go:22]: Not sure why a set is being used. Where are we building the set with available types?

File: pkg/blockstorage/azure/client.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/client.go:127]: Is there a typo here ? Should this be using blockstorage.AzureCientID instead of blockstorage.AzureTenantID ?

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:44]: Lets simplify this to 

`return  config[blockstorage.AzureTenantID]!= "" && config[blockstorage.AzureCientID] != "" && config[blockstorage.AzureClentSecret] != ""`

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:54]: Similar to the comment above lets simplify this to 

`return config[blockstorage.AzureTenantID] == "" && config[blockstorage.AzureClientID] != " && config[blockstorage.AzureClientSecret] == "" `

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:49]: typo AzureCientID -> AzureClientID

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:50]: typo AzureClentSecret -> AzureClientSecret

File: pkg/blockstorage/azure/client.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/client.go:106]: Please return `nil, errors.New.....` .

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:22]: Why does this have only the type for ManagedIdentity ?  I noticed that you have a function named getClientCredsAuthorizer which gives the impression that we do support something other than ManagedIdentity. But then  this function isCredTypeSupported only returns true if the type is CredTypeManagedIdentity .

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:19]: I dont see these two constants used in this file. Lets not include it if they are not used.

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:59]: Curious why you need another field of type CredsValidity when the error would have been sufficient ? My recommendation is to create new errors and export them in case the caller wants to look for specific errors/conditions.  With the current model of using CredsValidity, the caller would have to check and use both the validity field and the error field. It is simpler to have the caller deal with only an error field. 
```
var ErrMSIUnsupported = errors.New("MSI Endpoint is not supported")
var ErrServicePrincipalTokenCreateFailed = ...
var ErrRefreshTokenFailed = ....

```

and use them in msIAuthenticator's authenticate . A nil error would indicate that authentication was successful. Only when error is not nil, the caller could check the exported errors from this package.

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:72]: Is there a typo here ? Should we use blockstorage.ClientID instead of tenant ID ?

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:49]: It's a typo from source: https://github.com/kanisterio/kanister/blob/71b59c968301913f440cb81950fd3966938be310/pkg/blockstorage/helpers.go#L33

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:50]: It's a typo from source: 
https://github.com/kanisterio/kanister/blob/71b59c968301913f440cb81950fd3966938be310/pkg/blockstorage/helpers.go#L34

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:72]: Indeed. Thank you

File: pkg/blockstorage/azure/client.go
Comment (leuyentran) [reply pkg/blockstorage/azure/client.go:127]: indeed. Thank you

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:50]: I can add a refactor commit

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:49]: I can add a refactor commit

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:59]: I am not able to think of a design at the moment with other priorities. Between the caller's ability to look for specific errors/conditions and returning errors from Azure APIs, I think it makes sense to prioritize the latter to give end-user precise information on failures. I will : 
1. Remove  `CredsValidity`
2. Keep error returning as is 
Therefore, a successful authentication from the caller side is `error == nil `, any error otherwise will be piped back up the stack

What do you think @onkarbhat

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go]: Because the interface **Authenticated()** only supported ManagedIdentity at that point and that was PR's intention. However, that would be too confusing and thus I had added `ClientCredsAuthorizer`

Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go]: It does not set MSI as default, just a list of 1 since `Authenticated()` was only added for MSI. However I think it's confusing to only have MSI and not ClientCreds, so ClientCreds is added:
```go
var availableCredsType = sets.NewString(CredTypeManagedIdentity, CredTypeClientSecret)
```

a `set` is used so I can use an interface such as `Has()` to quickly search for supported credential types. When more credentials type is added in the future, only the `set` initialization needs to be extended

File: pkg/blockstorage/azure/auth.go
Comment (ihcsim) [inline pkg/blockstorage/azure/auth.go:6]: Looks like this package will be [deprecated next year](https://pkg.go.dev/github.com/Azure/go-autorest/autorest/adal@v0.9.21#section-readme). Can we look into using [`azidentity`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity) instead?

Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:6]: The purpose of interface `Authenticate()` is to validate Azure creds as a one-of action. These creds are then passed on to an `Authorizer()` that later authenticates and return a token for every Azure API calls that interact with Azure resources. 
And for `Authorizer()`, we are still using `adal` indirectly as part of  `github.com/Azure/go-autorest/autorest/azure/auth` 

Furthermore, `azidentity` and `adal` differs slightly in their credential definitions and workflows. So I'd prefer not to use `azidentity` for `Authenticate()` and `adal` for `Authorizer()` in case we don't completely understand the nuances in expectations.  

Instead, I had abstracted out the logics as much as I can, which should make APIs swapping easier during migration :)

**Review by ihcsim — COMMENTED** [review]
Code change LGTM. I will leave it to @onkarbhat and @bathina2 to approve.

**Comment by bathina2** [comment]
The changes look good. However, I do have some design opinions. From the caller side I would expect something like - 
```
authenticator, err := GetAuthenticator
if err != nil {
    //spit out err
}
err = authenticator.Authenticate()
```
Let me know if you have time to discuss tomorrow

File: pkg/blockstorage/azure/auth_tests.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth_tests.go:31]: This test is meant for testing the following definition and function. The definition of supported is if credential type is part of `availableCredsType` - which can be extended in the future

Test for `CredTypeClientSecret` is added 
```go
var availableCredsType = sets.NewString(CredTypeManagedIdentity, CredTypeClientSecret)

func isCredTypeSupported(credType string) bool {
	return availableCredsType.Has(credType)
}
```

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:67]: The caller, i.e. `Authenticate()` directly return results of these methods. Therefore,  the explicitly here is meant for code readability for future developers to understand: `err==nil` is considered passing authentication.

File: pkg/blockstorage/azure/auth_tests.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth_tests.go]: After discussion with @bathina2 , this function is now removed

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:6]: https://github.com/kanisterio/kanister/issues/1621

File: pkg/blockstorage/azure/auth.go
Comment (bathina2) [inline pkg/blockstorage/azure/auth.go:45]: nit: consider renaming to msiAzureAuthenticator

File: pkg/blockstorage/azure/auth.go
Comment (bathina2) [inline pkg/blockstorage/azure/auth.go:69]: nit: consider renaming to clientSecretAzureAuthenitcator.

**Review by bathina2 — APPROVED** [review]
Looks good, Minor nits about naming. Might be good to be more explicit, but not a big deal

File: pkg/blockstorage/azure/auth.go
Comment (leuyentran) [reply pkg/blockstorage/azure/auth.go:69]: Since the package name is `azure' , I think it is okay to leave the naming as is. If they need to be public for some reason in the future, it would still be `azure.ClientSecretAuthenticator`

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [reply pkg/blockstorage/azure/auth.go]: Sounds good to me. Thank you Le

File: pkg/blockstorage/azure/auth.go
Comment (onkarbhat) [inline pkg/blockstorage/azure/auth.go:33]: Please add comments for exported methods/structs in this PR.

**Review by onkarbhat — COMMENTED** [review]
LGTM . Posting it after merge.

**Pull request merged**
